consensus/hotstuff/signature/packager.go

```diff
@@ -1,26 +1,238 @@
+			return nil, fmt.Errorf("unknown signer ID (%v) at the given block (%v): %w",
+				signerID, blockID, signature.ErrInvalidFormat)
+		}
+
+		if sigType == hotstuff.SigTypeStaking {
+			stakingSigners = append(stakingSigners, signerID)
+		} else if sigType == hotstuff.SigTypeRandomBeacon {
+			randomBeaconSigners = append(randomBeaconSigners, signerID)
+		} else {
+			return nil, fmt.Errorf("unknown sigType %v, %w", sigType, signature.ErrInvalidFormat)
+		}
+	}
+
+	return &hotstuff.BlockSignatureData{
+		StakingSigners:               stakingSigners,
+		RandomBeaconSigners:          randomBeaconSigners,
+		AggregatedStakingSig:         data.AggregatedStakingSig,
+		AggregatedRandomBeaconSig:    data.AggregatedRandomBeaconSig,
+		ReconstructedRandomBeaconSig: data.RandomBeacon,
+	}, nil
+}
+
+// serializeToBytes encodes the given sigTypes into a bit vector.
+// We append tailing `0`s to the vector to represent it as bytes.
+func serializeToBytes(sigTypes []hotstuff.SigType) ([]byte, error) {
+	bytes := make([]byte, 0)
+	// a sig type can be converted into one bit.
+	// so every 8 sig types will fit into one byte.
+	// iterate through every 8 sig types, for each sig type in each
+	// group, convert into each bit, and fill the bit into the byte.
+	// the remaining unfilled bits in the last byte will be 0
+	for byt := 0; byt < len(sigTypes); byt += 8 {
+		b := byte(0)
+		offset := 7
+		for pos := byt; pos < byt+8 && pos < len(sigTypes); pos++ {
+			sigType := sigTypes[pos]
+
+			if !sigType.Valid() {
+				return nil, fmt.Errorf("invalid sig type: %v at pos %v", sigType, pos)
+			}
+
+			b ^= (byte(sigType) << offset)
+			offset--
+		}
+		bytes = append(bytes, b)
+	}
+	return bytes, nil
+}
+
+// deserializeFromBytes decodes the sig types from the given bit vector
+// - serialized: bit-vector, one bit for each signer (tailing `0`s appended to make full bytes)
+// - count: the total number of sig types to be deserialized from the given bytes
+// It returns:
+// - (sigTypes, nil) if successfully deserialized sig types
+// - (nil, signature.ErrInvalidFormat) if the number of serialized bytes doesn't match the given number of sig types
+// - (nil, signature.ErrInvalidFormat) if the remaining bits in the last byte are not all 0s
+func deserializeFromBytes(serialized []byte, count int) ([]hotstuff.SigType, error) {
+	types := make([]hotstuff.SigType, 0, count)
+
+	// validate the length of serialized
+	// it must have enough bytes to fit the `count` number of bits
+	totalBytes := count / 8
+	if count%8 > 0 {
+		totalBytes++
+	}
+
+	if len(serialized) != totalBytes {
+		return nil, fmt.Errorf("mismatching bytes for serialized sig types, total signers: %v"+
+			", expect bytes: %v, actual bytes: %v, %w", count, totalBytes, len(serialized), signature.ErrInvalidFormat)
+	}
+
+	// parse each bit in the bit-vector, bit 0 is SigTypeStaking, bit 1 is SigTypeRandomBeacon
+	for i := 0; i < count; i++ {
+		byt := serialized[i/8]
+		posMask := byte((1 << 7) >> (i % 8))
```

Comment:
here we first shift left by 7 and then shift right by `i % 8`. Instead of the forth and back, we can compute the shift upfront:
```suggestion
		offset := 7 - (i % 8)
		posMask := byte(1 << offset)
```